src/DataStructures/LinkedMessageQueue.hpp

```diff
@@ -0,0 +1,178 @@
+class LinkedMessageQueue<Id, tmpl::list<QueueTags...>> {
+  static_assert(sizeof...(QueueTags) > 0,
+                "Must have at least one message source.");
+
+ public:
+  using IdType = Id;
+
+  /// Insert data into a given queue at a given ID.  All queues must
+  /// receive data with the same collection of \p id_and_previous, but
+  /// are not required to receive them in the same order.
+  template <typename Tag>
+  void insert(const LinkedMessageId<Id>& id_and_previous,
+              typename Tag::type message) noexcept;
+
+  /// The next ID in the received sequence, if all queues have
+  /// received messages at that ID.
+  std::optional<Id> ready_id() const noexcept;
+
+  /// All the messages received at the time indicated by `ready_id()`.
+  /// It is an error to call this function if `ready_id()` would
+  /// return an empty optional.  This function removes the stored
+  /// messages from the queues, advancing the internal sequence ID.
+  tuples::TaggedTuple<QueueTags...> extract() noexcept;
+
+  // NOLINTNEXTLINE(google-runtime-references)
+  void pup(PUP::er& p) noexcept {
+    p | previous_id_;
+    p | messages_;
+  }
+
+ private:
+  template <typename Tag>
+  struct Optional {
+    using type = std::optional<typename Tag::type>;
+  };
+
+  using OptionalTuple = tuples::TaggedTuple<Optional<QueueTags>...>;
+
+  std::optional<Id> previous_id_{};
+  std::unordered_map<std::optional<Id>, std::pair<Id, OptionalTuple>>
+      messages_{};
+};
+
+template <typename Id, typename... QueueTags>
+template <typename Tag>
+void LinkedMessageQueue<Id, tmpl::list<QueueTags...>>::insert(
+    const LinkedMessageId<Id>& id_and_previous,
+    typename Tag::type message) noexcept {
+  static_assert((... or std::is_same_v<Tag, QueueTags>), "Unrecognized tag.");
+  const auto entry =
+      messages_
+          .insert({id_and_previous.previous,
+                   std::pair{id_and_previous.id, OptionalTuple{}}})
+          .first;
+  auto& value = entry->second;
+  auto& [id, tuple] = value;
+  ASSERT(id_and_previous.id == id,
+         "Received messages with different ids (" << id << " and "
+         << id_and_previous.id << ") but the same previous id ("
+         << id_and_previous.previous << ").");
+  ASSERT(not tuples::get<Optional<Tag>>(tuple).has_value(),
+         "Received duplicate messages at id " << id << " and previous "
+         << id_and_previous.previous << ".");
+  tuples::get<Optional<Tag>>(tuple) = std::move(message);
+}
+
+template <typename Id, typename... QueueTags>
+std::optional<Id> LinkedMessageQueue<Id, tmpl::list<QueueTags...>>::ready_id()
+    const noexcept {
+  const auto current_entry = messages_.find(previous_id_);
+  if (current_entry == messages_.end()) {
+    return {};
+  }
+  const auto& current_value = current_entry->second;
+  const auto& [current_id, current_messages] = current_value;
+  if ((... and
+       tuples::get<Optional<QueueTags>>(current_messages).has_value())) {
```

Comment:
Neat C++17 feature called a [fold expression](https://en.cppreference.com/w/cpp/language/fold).  In this case it is taking each entry in the `QueueTags` pack, evaluating the expression here with it, and then `and`ing all the results.  There's another one in the next function that uses `,` instead of `and`.

Reply:
😮 Oh that's cool!